keps/sig-network/3866-nftables-proxy/README.md

```diff
@@ -0,0 +1,1841 @@
+
+  3. The Cilium approach: Intercept the connect(2) call with eBPF and
+     rewrite the destination IP there, so that the network stack never
+     actually sees a packet with destination `127.0.0.1` / `::1`. (As
+     in the `userspace` kube-proxy case, this is not a special-case
+     for localhost, it's just how Cilium does service proxying.)
+
+  4. If you control the client, you can explicitly bind the socket to
+     `127.0.0.1` / `::1` before connecting. (I'm not sure why this
+     works since the packet still eventually gets routed off `lo`.) It
+     doesn't seem to be possible to "spoof" this after the socket is
+     created, though as with the previous case, you could do this by
+     intercepting syscalls with eBPF.
+
+In discussions about this feature, only one real use case has been
+presented: it allows you to run a docker registry in a pod and then
+have nodes use a NodePort service via `127.0.0.1` to access that
+registry. Docker treats `127.0.0.1` as an "insecure registry" by
+default (though containerd and cri-o do not) and so does not require
+TLS authentication in this case; using any other IP would require
+setting up TLS certificates, making the deployment more complicated.
+(In other words, this is basically an intentional exploitation of the
+security hole that CVE-2020-8558 warns about: enabling
+`route_localnet` may allow someone to access a service that doesn't
+require authentication because it assumed it was only accessible to
+localhost.)
+
+In all other cases, it is generally possible (though not always
+convenient) to just rewrite things to use the node IP rather than
+localhost (or to use a ClusterIP rather than a NodePort). Indeed,
+since localhost NodePorts do not work with `ipvs` mode or with IPv6,
+many places that used to use NodePorts on `127.0.0.1` have already
+been rewritten to not do so (eg [contiv/vpp#1434]).
+
+So:
+
+  - There is no way to make IPv6 localhost NodePorts work with a
+    NAT-based solution.
+
+  - The way to make IPv4 localhost NodePorts work with NAT introduces
+    a security hole, and we don't necessarily have a fully-generic way
+    to mitigate it.
+
+  - The only commonly-argued-for use case for the feature involves
+    deploying a service in a configuration which its own documentation
+    describes as insecure and "only appropriate for testing".
+
+      - The use case in question works by default against cri-dockerd
+        but not against containerd or cri-o with their default
+        configurations.
+
+      - cri-dockerd, containerd, and cri-o all allow additional
+        "insecure registry" IPs/CIDRs to be configured, so an
+        administrator could configure them to allow non-TLS image
+        pulling against a ClusterIP.
+
+Given this, I think we should not try to support localhost NodePorts
+in the `nftables` backend.
+
+```
+<<[UNRESOLVED dnat-but-no-route_localnet ]>>
+
+As a possible compromise, we could make the `nftables` backend create
```

Comment:
let's keep this in our pocket, in case a real use-case emerges, but not implement until then?